### ISSUE TITLE
Comparison against nullable struct incorrectly constrains SQL with != null in Select 

### ISSUE THREAD

**Issue by mqudsi** [dotnet/efcore#19454]
When `SelectMany` is used against a navigation property collection representing a join against another table and `DefaultIfEmpty()` is used to coerce a `LEFT JOIN` query, an incorrect additional constraint is added subsequent `Where` clauses that incorrectly narrows down the database results to rows with a non-null comparison target, although the left join is intended to produce a potentially null value.

For example, with `Foo` where `Foo.BarCollection` is a collection of `Bar` from table `Bar`:

```csharp
class Foo {
    ...
    List<Bar> BarCollection = new List<Bar>(); // nullable reference types
}

class Bar {
    ...
    NonNullableStruct NotNull, // with a db converter to-and-from string
```

In C#, an actual query:

```csharp
Ulid? graderId = null;
var response = await _db.Responses
    .Where(response => response.Id == responseId)
    .SelectMany(response => response.Gradings.DefaultIfEmpty(), (response, gradedSurvey) =>
    new
    {
        Response = response,
        GraderId = gradedSurvey.GraderId,
    })
    .Where(join => join.GraderId != graderId)
    .Select(join => join.Response)
    .Include(response => response.Survey.Questions)
        .ThenInclude(question => question.Answers)
    .OrderBy(r => r.Id)
    .FirstOrDefaultAsync(cancel)
    .ConfigureAwait(false);
```

here `Responses` is `Foo`, `Gradings` is `Bar`, `Bar.Survey.GraderId` is NonNullableStruct but `Bar.Survey` is another navigation property to a different table (e.g. nullable).

This expands to the following SQL(ite) at runtime:

```sql
SELECT "t"."id", "t"."date", "t"."gradeId", "t"."hash", "t"."path", "t"."subjectId", "t"."surveyId", "t"."teacherId", "t"."id0", "t"."date0", "t"."jpegPath", "t"."name", "t"."pdfPath", "t0"."id", "t0"."index", "t0"."name", "t0"."surveyId", "t0"."text", "t0"."id0", "t0"."index0", "t0"."name0", "t0"."questionId", "t0"."text0"
FROM (
    SELECT "r"."id", "r"."date", "r"."gradeId", "r"."hash", "r"."path", "r"."subjectId", "r"."surveyId", "r"."teacherId", "s"."id" AS "id0", "s"."date" AS "date0", "s"."jpegPath", "s"."name", "s"."pdfPath"
    FROM "responses" AS "r"
    LEFT JOIN "gradedSurveys" AS "g" ON "r"."id" = "g"."responseId"
    INNER JOIN "surveys" AS "s" ON "r"."surveyId" = "s"."id"
    WHERE ("r"."id" = @__responseId_0) AND "g"."graderId" IS NOT NULL
    ORDER BY "r"."id"
    LIMIT 1
) AS "t"
LEFT JOIN (
    SELECT "q"."id", "q"."index", "q"."name", "q"."surveyId", "q"."text", "a"."id" AS "id0", "a"."index" AS "index0", "a"."name" AS "name0", "a"."questionId", "a"."text" AS "text0"
    FROM "questions" AS "q"
    LEFT JOIN "answers" AS "a" ON "q"."id" = "a"."questionId"
) AS "t0" ON "t"."id0" = "t0"."surveyId"
ORDER BY "t"."id", "t"."id0", "t0"."id", "t0"."id0"
```

As you can see, the constraint is narrowed to records where the potentially null (as a result of an left join) equality target `g.graderId` is unnecessarily constrained with an `IS NOT NULL`.

When the object return in the join comparison linq expression is coerced to a nullable struct, as below, an exception is thrown at runtime:

```csharp
.SelectMany(response => response.Gradings.DefaultIfEmpty(), (response, gradedSurvey) =>
new
{
    Response = response,
    GraderId = (Ulid?) gradedSurvey.GraderId,
})
.Where(join => join.GraderId != graderId)
```

tl;dr after a LEFT JOIN against a table containing a non-nullable property, the SQL is coerced to return a non-null values from the resulting table rather than `Nullable<NonNullable>`, even though a comparison with a `Nullable<NonNullable>` is performed in the LINQ expression.

---

I assume there is an incorrect optimization (and probably not affecting only SQLite, although I did not have chance to test) that assumes a condition in `SelectMany` is intended to *match* rather than exclude a record (after all, this is `SelectMany` rather than `Join`).

### Additional context

Microsoft.Data.Sqlite version:  Core 3.1.0
Target framework: .NET Core 3.1
Operating system: Windows 10 x64 build 18362

**Comment by smitpatel** [comment]
Need to test this. Few things
- Include after Select is working which I would expect to throw
- The where condition is != but == in query.

**Comment by ajcvickers** [comment]
@smitpatel Do you think we need a full repro?

**Comment by smitpatel** [comment]
We can attempt to see what SQL we generate and if it matches out expectation for given shape of LINQ query.

**Comment by mqudsi** [comment]
This is an unnecessarily convoluted example, it boils down to an issue with the coercion of converted struct types when nullable. I'll refile with a much more succinct example.
